From b95028e468f73528146d57e68ffca71aa47b4cbe Mon Sep 17 00:00:00 2001
From: Andy Lutomirski <luto@kernel.org>
Date: Fri, 9 Dec 2016 10:24:07 -0800
Subject: [PATCH] x86/microcode: Replace sync_core() with cpuid_eax()

The Intel microcode driver is using sync_core() to mean "do CPUID
with EAX=1".

Signed-off-by: Andy Lutomirski <luto@kernel.org>
Acked-by: Borislav Petkov <bp@alien8.de>
[Linux commit 484d0e5c7943644cc46e7308a8f9d83be598f2b9]
[Ported to Xen]
Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Acked-by: Jan Beulich <jbeulich@suse.com>
diff --git a/xen/arch/x86/microcode_intel.c b/xen/arch/x86/microcode_intel.c
index 56b07707e8..617518b223 100644
--- a/xen/arch/x86/microcode_intel.c
+++ b/xen/arch/x86/microcode_intel.c
@@ -116,8 +116,9 @@ static int collect_cpu_info(unsigned int cpu_num, struct cpu_signature *csig)
     }
 
     wrmsrl(MSR_IA32_UCODE_REV, 0x0ULL);
-    /* see notes above for revision 1.07.  Apparent chip bug */
-    sync_core();
+    /* As documented in the SDM: Do a CPUID 1 here */
+    cpuid_eax(1);
+
     /* get the current revision from MSR 0x8B */
     rdmsrl(MSR_IA32_UCODE_REV, msr_content);
     csig->rev = (uint32_t)(msr_content >> 32);
@@ -297,8 +298,8 @@ static int apply_microcode(unsigned int cpu)
     wrmsrl(MSR_IA32_UCODE_WRITE, (unsigned long)uci->mc.mc_intel->bits);
     wrmsrl(MSR_IA32_UCODE_REV, 0x0ULL);
 
-    /* see notes above for revision 1.07.  Apparent chip bug */
-    sync_core();
+    /* As documented in the SDM: Do a CPUID 1 here */
+    cpuid_eax(1);
 
     /* get the current revision from MSR 0x8B */
     rdmsrl(MSR_IA32_UCODE_REV, msr_content);
diff --git a/xen/arch/x86/traps.c b/xen/arch/x86/traps.c
index 22c9310583..b352e766ad 100644
--- a/xen/arch/x86/traps.c
+++ b/xen/arch/x86/traps.c
@@ -3064,7 +3064,8 @@ static int emulate_privileged_op(struct cpu_user_regs *regs)
             {
                 if ( wrmsr_safe(MSR_IA32_UCODE_REV, 0) )
                     goto fail;
-                sync_core();
+                /* As documented in the SDM: Do a CPUID 1 here */
+                cpuid_eax(1);
             }
             goto rdmsr_normal;
         case MSR_IA32_MISC_ENABLE:
